util/src/RawMemoryPool.cpp

```diff
@@ -0,0 +1,148 @@
+  numOfAllocatedChunks_++;
+  metrics_.increaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::increaseNumOfAvailableChunks() {
+  numOfAvailableChunks_++;
+  metrics_.increaseAvailableChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAllocatedChunks() {
+  numOfAllocatedChunks_--;
+  metrics_.decreaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAvailableChunks() {
+  numOfAvailableChunks_--;
+  metrics_.decreaseAvailableChunksNum();
+}
+
+char* RawMemoryPool::allocateChunk() {
+  char* chunk = nullptr;
+  if (numOfAllocatedChunks_ < maxChunksNum_) {
+    chunk = new char[chunkSize_];
+    increaseNumOfAllocatedChunks();
+    increaseNumOfAvailableChunks();
+  }
+  if (chunk)
+    LOG_DEBUG(logger(), "A chunk has been allocated" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+  else
+    LOG_WARN(logger(),
+             "The pool size has reached the maximum, wait for a chunk to become available" << KVLOG(maxChunksNum_));
+  return chunk;
+}
+
+void RawMemoryPool::deleteChunk(char*& chunk) {
+  delete[] chunk;
+  decreaseNumOfAllocatedChunks();
+  LOG_DEBUG(logger(), "A chunk has been deleted" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+}
+
+void RawMemoryPool::returnChunkToThePool(char* chunk) {
+  pool_->push(chunk);
+  increaseNumOfAvailableChunks();
+  LOG_DEBUG(logger(), "A chunk has been returned to the pool" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+  waitForAvailChunkCond_.notify_one();
+}
+
+bool RawMemoryPool::isPoolPruningRequired() {
+  return (((numOfAllocatedChunks_ > initialChunksNum_) &&
+           (numOfAvailableChunks_ >= numOfAllocatedChunks_ * maxAvailableChunksPercentage_)));
+}
+
+char* RawMemoryPool::getChunk() {
+  ConcordAssert(pool_.get() != nullptr);
+  char* chunk = nullptr;
+  if (!pool_->pop(chunk)) {
+    // No available chunks => allocate a new one, if permitted
+    chunk = allocateChunk();
+    if (!chunk) {
+      // Pool size limit has been reached; wait until some chunk gets released
+      unique_lock<mutex> lock(waitForAvailChunkLock_);
+      while (!stopWorking_ && !chunk) {
+        waitForAvailChunkCond_.wait_until(lock, steady_clock::now() + milliseconds(WAIT_TIMEOUT_MILLI));
+        pool_->pop(chunk);
+      }
+    }
+  }
+  if (chunk) {
+    decreaseNumOfAvailableChunks();
+    LOG_DEBUG(logger(), "A chunk has been taken" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
+  }
+  return chunk;
+}
+
+void RawMemoryPool::returnChunk(char* chunk) {
+  if (numOfAvailableChunks_ >= maxChunksNum_) {
+    stringstream err;
+    err << "Returned chunk overflows the pool size limit: numOfAvailableChunks_=" << numOfAvailableChunks_
+        << ", maxChunksNum_=" << maxChunksNum_;
+    LOG_ERROR(logger(), KVLOG(err.str()));
+    throw std::runtime_error(__PRETTY_FUNCTION__ + err.str());
+  }
+  if (!isPoolPruningRequired())
+    returnChunkToThePool(chunk);
+  else  // Too many chunks are not in use => delete
+    deleteChunk(chunk);
```

Comment:
Yes, but I intentionally don't like to take strong decisions based on just one check (maybe we are going to have a peak starting from the next request?). Besides that, it could be time-consuming doing many delete operations before returning the chunk to the caller.